server/sonar-process-monitor/src/main/java/org/sonar/application/process/JavaProcessLauncherImpl.java

```diff
@@ -26,90 +26,88 @@
 
-  private final Timeouts timeouts;
   private final File tempDir;
   private final AllProcessesCommands allProcessesCommands;
+  private final Supplier<SystemProcessBuilder> processBuilderSupplier;
 
-  JavaProcessLauncher(Timeouts timeouts, File tempDir) {
-    this.timeouts = timeouts;
+  public JavaProcessLauncherImpl(File tempDir) {
+    this(tempDir, new AllProcessesCommands(tempDir), SystemProcessBuilder::new);
+  }
+
+  JavaProcessLauncherImpl(File tempDir, AllProcessesCommands allProcessesCommands, Supplier<SystemProcessBuilder> processBuilderSupplier) {
     this.tempDir = tempDir;
-    this.allProcessesCommands = new AllProcessesCommands(tempDir);
+    this.allProcessesCommands = allProcessesCommands;
+    this.processBuilderSupplier = processBuilderSupplier;
   }
 
+  @Override
   public void close() {
     allProcessesCommands.close();
   }
 
-  ProcessRef launch(JavaCommand command) {
+  @Override
+  public ProcessMonitor launch(JavaCommand javaCommand) {
     Process process = null;
+    ProcessCommands commands;
     try {
-      ProcessCommands commands = allProcessesCommands.createAfterClean(command.getProcessId().getIpcIndex());
+      commands = allProcessesCommands.createAfterClean(javaCommand.getProcessId().getIpcIndex());
 
-      ProcessBuilder processBuilder = create(command);
-      LoggerFactory.getLogger(getClass()).info("Launch process[{}]: {}",
-        command.getProcessId().getKey(), StringUtils.join(processBuilder.command(), " "));
+      SystemProcessBuilder processBuilder = create(javaCommand);
+      LOG.info("Launch process[{}]: {}", javaCommand.getProcessId().getKey(), String.join(" ", processBuilder.command()));
       process = processBuilder.start();
-      StreamGobbler inputGobbler = new StreamGobbler(process.getInputStream(), command.getProcessId().getKey());
-      inputGobbler.start();
-
-      return new ProcessRef(command.getProcessId().getKey(), commands, process, inputGobbler);
+      return new ProcessMonitorImpl(process, commands);
 
     } catch (Exception e) {
       // just in case
-      ProcessUtils.sendKillSignal(process);
-      throw new IllegalStateException("Fail to launch [" + command.getProcessId().getKey() + "]", e);
+      if (process != null) {
+        process.destroyForcibly();
+      }
+      throw new IllegalStateException(format("Fail to launch process [%s]", javaCommand.getProcessId().getKey()), e);
     }
   }
 
-  private ProcessBuilder create(JavaCommand javaCommand) {
+  private SystemProcessBuilder create(JavaCommand javaCommand) {
     List<String> commands = new ArrayList<>();
     commands.add(buildJavaPath());
     commands.addAll(javaCommand.getJavaOptions());
     // TODO warning - does it work if temp dir contains a whitespace ?
-    commands.add(String.format("-Djava.io.tmpdir=%s", tempDir.getAbsolutePath()));
-    commands.add(getJmxAgentCommand());
+    // TODO move to JavaCommandFactory ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)